ports/esp32/modmachine.c

```diff
@@ -254,6 +255,40 @@ static mp_obj_t machine_wake_reason(size_t n_args, const mp_obj_t *pos_args, mp_
 }
 static MP_DEFINE_CONST_FUN_OBJ_KW(machine_wake_reason_obj, 0,  machine_wake_reason);
 
+static mp_obj_t machine_wake_pins(void) {
+    uint64_t status = 0;
+    int len, index;
+    mp_obj_t *tuple = NULL;
+
+    // There will be only one wake-up source, so it is OK to logically OR all the
+    // wake-up source statuses.
+    #if SOC_GPIO_SUPPORT_DEEPSLEEP_WAKEUP && SOC_DEEP_SLEEP_SUPPORTED
+    status |= esp_sleep_get_gpio_wakeup_status();
+    #endif
+
+    #if SOC_PM_SUPPORT_EXT1_WAKEUP && SOC_RTCIO_PIN_COUNT > 0
+    status |= esp_sleep_get_ext1_wakeup_status();
+    #endif
+
+    // Only a few (~8) pins might cause wakeup.
+    // Therefore, we don't allocate 64*4 = 256 bytes on the stack and calculate the
+    // required space in a first pass.
+    for (index = 0, len = 0; index < 64; index++) {
+        len += (status & ((uint64_t)1 << index)) ? 1 : 0;
+    }
+    if (len) {
+        tuple = alloca(len * sizeof(*tuple));
```

Comment:
Rather than using `alloca` here suggest doing something like:

```c
#include "py/objtuple.h"

mp_obj_tuple_t *tuple = MP_OBJ_TO_PTR(mp_obj_new_tuple(len, NULL));
[...]
   tuple->items[i] = MP_OBJ_NEW_SMALL_INT(index)
[...]
return MP_OBJ_FROM_PTR(tuple)
```